Sources/ApolloCodegenLib/Frontend/CompilationResult.swift

```diff
@@ -0,0 +1,184 @@
+  }
+  
+  public struct VariableDefinition: Decodable {
+    var name: String
+    var type: GraphQLType
+    var defaultValue: GraphQLValue?
+  }
+  
+  public struct FragmentDefinition: Decodable {
+    var fragmentName: String
+    var type: GraphQLCompositeType
+    var selectionSet: SelectionSet
+    
+    var source: String
+    var filePath: String
+  }
+  
+  public struct SelectionSet: Decodable {
+    var possibleTypes: [GraphQLObjectType]
+    var selections: [Selection]
+  }
+  
+  public enum Selection: Decodable {
+    case field(Field)
+    case fragmentSpread(FragmentSpread)
+    case typeCondition(TypeCondition)
+    case booleanCondition(BooleanCondition)
+    
+    enum CodingKeys: CodingKey {
+      case kind
+    }
+    
+    public init(from decoder: Decoder) throws {
+      let container = try decoder.container(keyedBy: CodingKeys.self)
+      let kind = try container.decode(String.self, forKey: .kind)
+      
+      switch kind {
+      case "Field":
+        self = .field(try Field(from: decoder))
+      case "FragmentSpread":
+        self = .fragmentSpread(try FragmentSpread(from: decoder))
+      case "TypeCondition":
+        self = .typeCondition(try TypeCondition(from: decoder))
+      case "BooleanCondition":
+        self = .booleanCondition(try BooleanCondition(from: decoder))
+      default:
+        throw DecodingError.dataCorruptedError(forKey: .kind, in: container, debugDescription: """
+          Unknown selection kind "\(kind)"
+          """)
+      }
+    }
+  }
+  
+  public struct Field: Decodable, CustomStringConvertible {
+    var name: String
+    var alias: String?
+    
+    var responseKey: String {
+      alias ?? name
+    }
+    
+    var arguments: [Argument]?
+    
+    var type: GraphQLType
+    
+    var selectionSet: SelectionSet?
+    
+    var deprecationReason: String?
+    
+    var isDeprecated: Bool {
+      return deprecationReason != nil
+    }
+    
+    public var description: String {
+      return "\(responseKey): \(type)"
+    }
+  }
+  
+  public struct Argument: Decodable {
+    var name: String
+    var value: GraphQLValue
+  }
+  
+  public struct FragmentSpread: Decodable {
+    var fragmentName: String
+    var selectionSet: SelectionSet
+  }
+  
+  public struct TypeCondition: Decodable {
+    var type: GraphQLCompositeType
+    var selectionSet: SelectionSet
+  }
+  
+  public struct BooleanCondition: Decodable {
+    var variableName: String
+    var isInverted: Bool
+    var selectionSet: SelectionSet
+    
+    enum CodingKeys: String, CodingKey {
+      case variableName
+      case isInverted = "inverted"
+      case selectionSet
+    }
+  }
+}
+
+// Type references
+
+enum GraphQLTypeReferenceError: Error {
+  case syntaxError(String)
+  case namedTypeNotFound(String)
+}
+
+extension CodingUserInfoKey {
+  /// Resolving type names requires access to a GraphQL schema during decoding.
+  static var graphQLSchema = CodingUserInfoKey(rawValue: "graphQLSchema")!
```

Comment:
The JSON we get for the compilation result includes type references as strings. When decoding we need to look up types by name in the schema, so that requires having the schema available.

Reply:
Right that's what the comment says 😃 - why do we need a key for it?

Reply:
That's how providing context during decoding works. `Decoder` defines `var userInfo: [CodingUserInfoKey : Any]` and it expects you to create your own `CodingUserInfoKey` to put data on there and access it.